Lib/test/test_signal.py

```diff
@@ -112,6 +154,19 @@ def test_issue9324(self):
         with self.assertRaises(ValueError):
             signal.signal(7, handler)
 
+    @unittest.skipUnless(sys.executable, "sys.executable required.")
+    def test_keyboard_interrupt_exit_code(self):
+        """KeyboardInterrupt triggers an exit using STATUS_CONTROL_C_EXIT."""
+        # I tried using this as the child code:
+        #  "import os,signal; os.kill(os.getpid(), signal.CTRL_C_EVENT)"
+        # but that caused the entire testsuite runner to receive the
+        # event and bail out due to an interrupt on CI systems.
+        process = subprocess.run(
```

Comment:
`kill(pgid, CTRL_C_EVENT)` calls WINAPI `GenerateConsoleCtrlEvent`. This takes a process group ID, i.e. the ID of a process created with `creationflags=subprocess.CREATE_NEW_PROCESS_GROUP`. You encountered a bug in the console. If it's not a group ID, but the target is a child of a process that's attached to the console, then it sends the event to all attached processes (i.e. group 0). 

If you switch back to using `kill`, this time with a new process group, then Ctrl+C has to be enabled first (via `SetConsoleCtrlHandler`) since it's disabled by default in a new group. Also, console control events are delivered on a new thread, so we have a synchronization problem. We can have the main thread wait in `time.sleep`, and move the `kill` call to a worker thread, which ensures the event is received while Python is still running. For example:

    process = subprocess.run([sys.executable, "-c", 
                "import os, signal, time, threading, ctypes;"
                "ctypes.windll.kernel32.SetConsoleCtrlHandler(None, False);"
                "threading.Timer(1, os.kill, (os.getpid(), signal.CTRL_C_EVENT)).start();"
                "time.sleep(5)"],
                creationflags=subprocess.CREATE_NEW_PROCESS_GROUP, 
                stderr=subprocess.PIPE)

Reply:
Thanks!  I figured it was related to some grouping behaviors like that.  I've updated the comment, but I'm not going to adopt such code here as that seems a lot more complicated and really overkill just for this test.  If you or any other Windows maintainers feel it is worthwhile to include, feel free to update the test in a future PR. :)